### PR TITLE
Enforcing database transactions with Rust

### PR DESCRIPTION
**Pull request by jgke**

### CODE REVIEW

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+## Enter Rust
+
+Rust is a relatively new programming language, with focus on speed and safety.
+Rust has the unique notion of 'ownership', which enables memory safety without
+relying on a garbage collector. Nowadays in most languages all variables are
+actually references, which are just thrown around and are either wildly mutable
+or completely immutable. Rust provides a way to write code where mutation is
+explicit and safe.
+
+In my opinion, the best benefit of the borrow checker is not the speedup it
```
Comment (msantala) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: Jumps in to talking about the benefits of the borrow checker without actually introducing it. Maybe add a few lines on how it relates to the concepts in the previous paragraph?

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+Let's start by defining them:
+
+```rust
+pub struct Connection(Box<postgres::Connection>);
+pub struct Transaction<'a>(Box<postgres::transaction::Transaction<'a>>);
+```
+
+These types contain exactly one member: the respective `postgres` connection
+type, with the crucial difference that `Connection` and `Transaction` do not
+derive the `Clone` trait used to acquire new copies of the connection.
```
Comment (msantala) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: Short explanation of traits could be helpful here.

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+```rust
+pub fn get_person<IGC: IntoGenericConnection>(db: IGC, name: &str) -> Option<Person> {
+    let conn = db.into_generic_connection();
+    conn.query("SELECT id, name FROM account WHERE name=$1", &[&name]).unwrap()
+        .into_iter()
+        .map(|row| Person { id: row.get(0), username: row.get(1) })
+        .next()
+}
+```
+
```
Comment (msantala) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: Maybe re-iterate the benefits before moving on to performance?

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  while still allowing single queries without transactions.
+tags:
+ - Database
+ - Rust
+---
+
+Consider this Clojure snippet:
+
+```clojure
+(defn fetch-thing [user-id thing-id]
```
Comment (tatut) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: nobody would write clojure like this... you need to pass the db in, not have it magically from the "environment"... I know this is a nitpicky thing

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+## Enter Rust
+
+Rust is a relatively new programming language, with focus on speed and safety.
+Rust has the unique notion of 'ownership', which enables memory safety without
+relying on a garbage collector. Nowadays in most languages all variables are
+actually references, which are just thrown around and are either wildly mutable
+or completely immutable. Rust provides a way to write code where mutation is
+explicit and safe.
+
+Rust implements this by providing an unique feature: the borrow checker.
```
Comment (tatut) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: "_a_ unique feature" I think

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+Essentially, the borrow checker is a compile-time feature which ensures that
+objects live long enough, and at the same time prevents unsafe concurrent access
+to variables. This is implemented by following a couple simple rules:
```
Comment (tatut) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: "a couple _of_ simple rules"

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+When writing generic code, generic type parameters can receive trait bounds to
+specify things that can be done with the types. For example, in the following
+function
+
+```rust
+fn debug_print<T: Debug>(t: T) { dbg!(t); }
+```
+
+the type T is required to implement the trait `Debug`, which allows turning the
+object into an programmer-readable (but not necessarily human-readable) format.
```
Comment (tatut) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: is this a joke about programmers != humans or do you mean "human readable" as in suitable for UI? I couldn't tell.

Comment (jgke) [reply _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: Both :)

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+The generic functions are compiled similarily as in C++: any calls to the
+function are specialized to the types, which technically can create binary
+bloat, but practically reduces it as the compiler has better options for
+optimization.
+
+## Modeling database queries around ownership
+
+So, we need to be able to
```
Comment (tatut) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: Stylistic issue, I would expand the three things as text and not have a numbered list here... ymmv.

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md (hunk 7 above)
Comment (jgke) [reply _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: For the record, there's actually a different trait (std::format::Display) which provides human-readable output.

File: _posts/2019-11-15-enforcing-database-transactions-with-rust.md [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+1. Make a single database query without a transaction
+2. Convert a database connection into a transaction
+3. Make multiple queries inside a transaction
+
+Sounds like while making a query without a transaction, the connection should
+refuse any additional queries. If the queries are made within a transaction,
+additional queries should be accepted. Let's sketch an API for it:
```
Comment (tatut) [inline _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: should the tx case mention that the transaction is cleaned up properly

Comment (jgke) [reply _posts/2019-11-15-enforcing-database-transactions-with-rust.md]: Nice catch, the tx is rolled back by default when dropped, I added some .commit()s to it.

**Pull request merged**